sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs

```diff
@@ -2429,6 +2431,188 @@ impl ty::TyExpression {
         })
     }
 
+    pub fn find_subfield_type(
+        module: &Module,
+        handler: &Handler,
+        engines: &Engines,
+        namespace: &Namespace,
+        base_name: &Ident,
+        projections: &[ty::ProjectionKind],
+    ) -> Result<(TypeId, TypeId), ErrorEmitted> {
+        let mut lexical_scope_opt = Some(module.current_lexical_scope());
+        while let Some(lexical_scope) = lexical_scope_opt {
+            let result = Self::find_subfield_type_helper(
+                lexical_scope,
+                handler,
+                engines,
+                namespace,
+                base_name,
+                projections,
+            )?;
+            if let Some(result) = result {
+                return Ok(result);
+            }
+            if let Some(parent_scope_id) = lexical_scope.parent {
+                lexical_scope_opt = module.get_lexical_scope(parent_scope_id);
+            } else {
+                lexical_scope_opt = None;
+            }
+        }
```

Comment:
The pattern 

```
let mut lexical_scope_opt = ...;
while let Some(lexical_scope) = lexical_scope_opt {
  ...
  if let Some(parent_scope_id) = lexical_scope.parent {
    lexical_scope_opt = module.get_lexical_scope(parent_scope_id);
  } else {
    lexical_scope_opt = None;
  }
}
```
is repeated in a couple of places, and seems to capture a structural aspect of hierarchical lexical scopes. Is there a way to abstract this pattern out (ideally so that it is located in `lexical_scope.rs`), so that searching through the scopes is kept separate from what we are searching for?